starlark/testdata/bytes.star

```diff
@@ -0,0 +1,131 @@
+assert.fails(lambda: bytes([300]),
+             "at index 0, 300 out of range .want value in unsigned 8-bit range")
+assert.fails(lambda: bytes([b"a"]),
+             "at index 0, got bytes, want int")
+assert.fails(lambda: bytes(1), "want string, bytes, or iterable of ints")
+
+# literals
+assert.eq(b"hello, 世界", hello)
+assert.eq(b"goodbye", goodbye)
+assert.eq(b"", empty)
+assert.eq(b"\t\n\u200D", nonprinting)
+assert.ne("abc", b"abc")
+assert.eq(b"\012\xff\u0400\U0001F63F", b"\n\xffЀ😿") # see scanner tests for more
+assert.eq(rb"\r\n\t", b"\\r\\n\\t") # raw
+
+# type
+assert.eq(type(hello), "bytes")
+
+# len
+assert.eq(len(hello), 13)
+assert.eq(len(goodbye), 7)
+assert.eq(len(empty), 0)
+
+# truth
+assert.true(hello)
+assert.true(goodbye)
+assert.true(not empty)
+
+# str(bytes) does UTF-8 to UTF-k transcoding.
+assert.eq(str(hello), "hello, 世界")
+assert.eq(str(hello[:-1]), "hello, 世��")  # incomplete UTF-8 encoding => U+FFFD
+assert.eq(str(goodbye), "goodbye")
+assert.eq(str(empty), "")
+assert.eq(str(nonprinting), "\t\n\u200d")
+assert.eq(str(b"\udc00"), "�") # unpaired surrogate => U+FFFD
+
+# reprq
+assert.eq(repr(hello), r'b"hello, 世界"')
+assert.eq(repr(hello[:-1]), r'b"hello, 世\xe7\x95"')  # (incomplete UTF-8 encoding )
+assert.eq(repr(goodbye), 'b"goodbye"')
+assert.eq(repr(empty), 'b""')
+assert.eq(repr(nonprinting), 'b"\\t\\n\\u200d"')
+
+# equality
+assert.eq(hello, hello)
+assert.ne(hello, goodbye)
+assert.eq(bytes("goodbye"), goodbye)
+
+# ordered comparison
+assert.lt(bytes("abc"), bytes("abd"))
+assert.lt(bytes("abc"), bytes("abcd"))
+assert.lt(bytes("\x7f"), bytes("\x80")) # bytes compare as uint8, not int8
+
+# bytes are dict-hashable
+dict = {hello: 1, goodbye: 2}
+dict[bytes("goodbye")] = 3
+assert.eq(len(dict), 2)
+assert.eq(dict[goodbye], 3)
+
+# indexing
+assert.eq(goodbye[0], bytes("g"))
+assert.eq(goodbye[-1], bytes("e"))
+assert.fails(lambda: goodbye[100], "out of range")
+
+# slicing
+assert.eq(goodbye[:4], bytes("good"))
+assert.eq(goodbye[4:], bytes("bye"))
+assert.eq(goodbye[::2], bytes("gobe"))
+assert.eq(goodbye[3:4], bytes("d"))  # special case: len=1
+assert.eq(goodbye[4:4], bytes(""))  # special case: len=0
+
+# ord
+assert.eq(ord(b"a"), 97)
+assert.fails(lambda: ord(b"ab"), "ord: bytes has length 2, want 1")
+assert.fails(lambda: ord(b""), "ord: bytes has length 0, want 1")
+
+# repeat (bytes * int)
+assert.eq(goodbye * 3, bytes("goodbyegoodbyegoodbye"))
+assert.eq(3 * goodbye, bytes("goodbyegoodbyegoodbye"))
+
+# elems() returns an iterable value over 1-byte substrings.
+assert.eq(type(hello.elems()), "bytes.elems")
+assert.eq(str(hello.elems()), "b\"hello, 世界\".elems()")
+assert.eq(list(hello.elems()),
+          [b"h", b"e", b"l", b"l", b"o", b",", b" ",
+           b"\xE4", b"\xB8", b"\x96", b"\xE7", b"\x95", b"\x8C"])
+assert.eq(list(goodbye.elems()),
+          [b"g", b"o", b"o", b"d", b"b", b"y", b"e"])
+assert.eq(list(empty.elems()), [])
+
+# TODO(adonovan);
```

Comment:
(New thread: GitHub won't let me comment on an old thread after a force-push)

> The only missing case now is how to create a 1-element string, or a string from a list of int values, analogous to bytes([int]). Any ideas?

> Take a look at this updated comment for a summary of the various operations.

For a 1-element string from an integer code point, `chr` should work. That can produce multiple UTF-k elements, depending on the implementation.

For a string from a list of int code points, I don't know of a way to do that in Python. `str([1, 2, 3])` should return literally "[1, 2, 3]"`. It looks like Python `bytes` has a `decode` method. Maybe we should implement that, only supporting `"utf-8"` (or no argument)?

Reply:
> For a 1-element string from an integer code point, chr should work. That can produce multiple UTF-k elements, depending on the implementation.

Right, chr(int) produces a 1-code point string of multiple UTF-k codes (elements). But how would one create a string of exactly one element?

> For a string from a list of int code points

That's easy enough, if inefficient: "".join(["%c" % cp for cp in codepoints]).

Reply:
See my comment below. For implementations with k=8, I think you need \xXX literals up to 0xFF. For k=16, I think it suffices to have \uXXXX and allow unpaired surrogates.

Reply:
\x escapes don't allow non-ASCII, and \u escapes don't allow unpaired surrogates, because that would be non-portable, as would the now-abandoned \X escape notation. Now, there is simply no literal syntax to denote a string with an invalid (non-text) encoding.  (In the Rust implementation, such values cannot even be represented.)